packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.ts

```diff
@@ -0,0 +1,148 @@
+import { OptionsGroup, OptionsUIType, OptionEditor, OptionUIComponentProps } from '../../types/panelOptions';
+import { PanelOptionsGroup } from '../PanelOptionsGroup/PanelOptionsGroup';
+import { Omit } from '../../types/utils';
+import React from 'react';
+import { BooleanOption } from './BooleanOption';
+import { IntegerOption } from './NumericInputOption';
+import { FieldDisplayOptions } from '../../utils/index';
+import { FieldDisplayEditor, FieldPropertiesEditor } from '../SingleStatShared/index';
+import { Threshold } from '../../types/index';
+import { ThresholdsEditor } from '../ThresholdsEditor/ThresholdsEditor';
+
+type GroupConfig<TConfig> = TConfig & {
+  component?: React.ComponentType<Omit<TConfig, 'component'>>;
+};
+
+interface UIModelBuilder<TModel> {
+  getUIModel(): TModel;
+}
+
+interface EditorOptions {
+  label?: string;
+  placeholder?: string;
+  description?: string;
+}
+
+interface EditorConfig<TValueType extends {}> extends EditorOptions {
+  component: React.ComponentType<OptionUIComponentProps<TValueType>>;
+}
+
+type InferOptionType<TOptions extends object, TKey extends keyof TOptions> = TOptions[TKey];
+type KeysMatching<T, V> = { [K in keyof T]: T[K] extends V ? K : never }[keyof T];
+class OptionEditorUIBuilder implements UIModelBuilder<OptionEditor<any, any>> {
+  private model: EditorConfig<any>;
+
+  constructor(config: EditorConfig<any>) {
+    this.model = config;
+  }
+
+  getUIModel = () => {
+    return {
+      type: OptionsUIType.Editor as OptionsUIType.Editor, // TODO: how to fix this,
+      editor: this.model as any,
+    };
+  };
+}
+
+export class OptionsGroupUIBuilder<TOptions extends {}, TConfig extends {} = {}>
```

Comment:
@torkelo @ryantxu that's the first version of the API for options ui schema building. No connection with yup schema yet.